### PR TITLE
[9.0] feat: introduce the `JobWrapperOfflineTemplate` for resources with no external connectivity

### PR DESCRIPTION
**Pull request by aldbr**
PR2 from https://github.com/DIRACGrid/DIRAC/issues/7459

- `JobWrapperLightTemplate`: specific template for resources with no external connectivity. Assumes that inputs are available and pre-processing has been done. Solely executes the payload and exits.
- A few adjustments in the `JobWrapperTemplate` covered by extensive tests.

BEGINRELEASENOTES
*WorkloadManagement
NEW: introduce JobWrapperOfflineTemplate for uses in systems without external connectivity
ENDRELEASENOTES

### CODE REVIEW

**Comment by aldbr** [comment]
During the latest Bild-dev meeting, @andresailer suggested adopting a more explicit name to replace `JobWrapperLightTemplate`. Do you have any suggestions?

Here is a list of potential names for your consideration:
- JobWrapperSimpleTemplate
- JobWrapperCoreTemplate
- JobWrapperBasicTemplate
- JobWrapperProcessorTemplate
- JobWrapperExecutionTemplate

I am leaning towards `JobWrapperProcessorTemplate`. What are your thoughts?

**Comment by andresailer** [comment]
JobWrapperOfflineTemplate, JobWrapperAirGappedTemplate

**Comment by aldbr** [comment]
`JobWrapperOfflineTemplate` sound much more explicit and simple, thanks @andresailer!

File: src/DIRAC/WorkloadManagementSystem/JobWrapper/JobWrapperOfflineTemplate.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    # Store the payload result
+    with open(arguments["Job"]["PayloadResults"], "w") as f:
+        json.dump(payloadResult, f)
+
+    # Generate the checksum of the files present in the current directory
+    checksums = {}
+    for file in os.listdir("."):
+        if os.path.isfile(file):
+            hash_md5 = hashlib.md5()
```
Comment (andresailer) [inline src/DIRAC/WorkloadManagementSystem/JobWrapper/JobWrapperOfflineTemplate.py]: In python 3.11 there is https://docs.python.org/3/library/hashlib.html#file-hashing

File: src/DIRAC/WorkloadManagementSystem/Utilities/Utils.py [hunk 2]
```diff
@@ -4,98 +4,53 @@
...
+    :param jobID: Job ID
+    :param jobParams: Job parameters
+    :param resourceParams: CE parameters
+    :param optimizerParams: Optimizer parameters
+    :param payloadParams: Payload parameters
+    :param extraOptions: Extra options to be passed to the job wrapper
+    :param wrapperPath: Path where the job wrapper will be created
+    :param rootLocation: Location where the job wrapper will be executed
+    :param defaultWrapperLocation: Location of the default job wrapper template
+    :param pythonPath: Path to the python executable
```
Comment (andresailer) [inline src/DIRAC/WorkloadManagementSystem/Utilities/Utils.py]: In the function signature the order is reversed for these two lines
```suggestion
    :param pythonPath: Path to the python executable
    :param defaultWrapperLocation: Location of the default job wrapper template
```

File: src/DIRAC/Resources/Computing/SingularityComputingElement.py [hunk 3]
```diff
@@ -255,20 +255,21 @@ def __createWorkArea(self, jobDesc=None, log=None, logLevel="INFO", proxy=None):
...
             optimizerParams=jobDesc.get("optimizerParams", {}),
+            pythonPath="python",
             log=log,
             logLevel=logLevel,
             extraOptions="" if self.__installDIRACInContainer else "/tmp/pilot.cfg",
         )
         if not result["OK"]:
             return result
-        wrapperPath = result["Value"]
+        wrapperPath = result["Value"]["JobExecutableRelocatedPath"]
```
Comment (andresailer) [inline src/DIRAC/Resources/Computing/SingularityComputingElement.py]: `JobExecutableRelocatedPath` seems to be an optional entry in the dictionary? is it guaranteed that this is always present here?

**Comment by fstagni** [comment]
https://github.com/DIRACGrid/DIRAC/pull/7570 would provide additional tests for this PR. Can you review that one and (once merged) rebase this PR?

**Comment by aldbr** [comment]
Thanks @andresailer for the review, I applied the changes.
@fstagni I rebased the PR.

I just have one issue with `readthedocs` but I am not sure what to do with this:

```
WARNING: Failed to import DIRAC.WorkloadManagementSystem.JobWrapper.JobWrapperOfflineTemplate.
Possible hints:
* AttributeError: module 'DIRAC.WorkloadManagementSystem.JobWrapper' has no attribute 'JobWrapperOfflineTemplate'
* SystemExit: 2
```

**Comment by andresailer** [comment]
I think you have to add this to the ignore list
https://github.com/DIRACGrid/DIRAC/blob/46adcc03afde8a06eb1d3ab765e1f5c910329097/docs/docs.conf#L35-L36

**Review by fstagni — DISMISSED** [review]

**Review by fstagni — APPROVED** [review]

**Comment by andresailer** [comment]
Doc still fails, I will try to reproduce locally.

**Comment by andresailer** [comment]
No, now it is that randomly occuring WorkflowXMLReader issue... I hit rebuild for the docs.

**Pull request merged**